daemon/container_operations.go

```diff
@@ -574,52 +574,86 @@ func hasUserDefinedIPAddress(ipamConfig *networktypes.EndpointIPAMConfig) bool {
 	return ipamConfig != nil && (len(ipamConfig.IPv4Address) > 0 || len(ipamConfig.IPv6Address) > 0)
 }
 
-// User specified ip address is acceptable only for networks with user specified subnets.
-func validateNetworkingConfig(n *libnetwork.Network, epConfig *networktypes.EndpointSettings) error {
-	if n == nil || epConfig == nil {
+// validateEndpointSettings checks whether the given epConfig is valid. The nw parameter might be nil as a container
+// can be created with a reference to a network that don't exist yet. In that case, only partial validation will be
+// done.
+func validateEndpointSettings(nw *libnetwork.Network, nwName string, epConfig *networktypes.EndpointSettings) error {
+	if epConfig == nil {
 		return nil
 	}
-	if !containertypes.NetworkMode(n.Name()).IsUserDefined() {
+
+	var errs []error
+	if !containertypes.NetworkMode(nwName).IsUserDefined() {
+		// On Linux, user specified IP address is accepted only for networks with user specified subnets.
 		if hasUserDefinedIPAddress(epConfig.IPAMConfig) && !enableIPOnPredefinedNetwork() {
-			return runconfig.ErrUnsupportedNetworkAndIP
+			errs = append(errs, runconfig.ErrUnsupportedNetworkAndIP)
 		}
 		if len(epConfig.Aliases) > 0 && !serviceDiscoveryOnDefaultNetwork() {
-			return runconfig.ErrUnsupportedNetworkAndAlias
+			errs = append(errs, runconfig.ErrUnsupportedNetworkAndAlias)
 		}
 	}
+
 	if !hasUserDefinedIPAddress(epConfig.IPAMConfig) {
+		return errors.Join(errs...)
+	}
+	if epConfig.IPAMConfig.IPv4Address != "" && net.ParseIP(epConfig.IPAMConfig.IPv4Address).To4() == nil {
+		errs = append(errs, fmt.Errorf("invalid IPv4 address: %s", epConfig.IPAMConfig.IPv4Address))
+	}
+	if epConfig.IPAMConfig.IPv6Address != "" {
+		n := net.ParseIP(epConfig.IPAMConfig.IPv6Address)
+		// if the address is an invalid network address (ParseIP == nil) or if it is
+		// an IPv4 address (To4() != nil), then it is an invalid IPv6 address
+		if n == nil || n.To4() != nil {
+			errs = append(errs, fmt.Errorf("invalid IPv6 address: %s", epConfig.IPAMConfig.IPv6Address))
+		}
+	}
+	for _, addr := range epConfig.IPAMConfig.LinkLocalIPs {
+		if net.ParseIP(addr) == nil {
+			errs = append(errs, fmt.Errorf("invalid link-local IP address %s", addr))
+		}
+	}
+
+	if nw == nil {
```

Comment:
This validation is called from 2 places;

In `Daemon.updateNetworkConfig()`;

```go
if err := validateEndpointSettings(n, n.Name(), endpointConfig); err != nil {
	return err
}
```

In `Daemon.validateNetworkingConfig()`;

```go
// The referenced network k might not exist when the container is created, so don't fail if that's the case.
// If any other error happen, it's not a validation error, so better early exit.
nw, err := daemon.FindNetwork(k)
if err != nil && !errdefs.IsNotFound(err) {
	return err
}

if err := validateEndpointSettings(nw, k, v); err != nil {
	errs = append(errs, fmt.Errorf("invalid endpoint config for network %s: %w", k, err))
}
```

- in both cases, I think we'd already barf out if the network is not to be found
- and, I think in both cases `nwName` is effectively equivalent to `nw.Name()` (although I'm not 100% sure for the second case; can it use network ID instead of name? does that matter for this case?)

So I'm wondering if we should continue to bail out early if there's no network (instead of continuing with other checks), as we probably would never reach this validation

Also slightly looking if we should split "heavy" and "lightweight" validation; e.g. validating for invalid / invalidly formatted IP-addresses is something that's relatively lightweight, whereas fetching the network (knowing libnetwork ...) may "not be". From that perspective we could potentially do a quick check on well-formed-ness of options (we should really look if we can do such parsing/conversions earlier, and get to a "strong-type" ASAP in the process, as we're sometimes parsing the same string multiple times), and only if we get that far, proceed with the heavier lifting.

Reply:
> in both cases, I think we'd already barf out if the network is not to be found

Nope, we actually accept a `ContainerCreate` request with a reference to a non-existant network:

https://github.com/moby/moby/blob/b227bc4c7caad4267e5c7af8cba8687a8b8ab6a8/daemon/container_operations.go#L577-L580

It's not a hypothetical case, it's how things are working right now :wink:  

```
$ docker network inspect foobar
[]
Error response from daemon: network foobar not found

$ docker container create --network foobar busybox
6697a506ee56c76d770218b49794255b8059a0f6deec836a0c26778379612b6a
```

> we should really look if we can do such parsing/conversions earlier, and get to a "strong-type" ASAP in the process

Yeah, I looked into this a bit already. I'd like to have `netip.*` instead of strings, as it'd remove the need of constantly parsing the same thing. But that's imply quite a bunch of work. I did had time to create the CLI PR that depends on this one, but it all have to do with the ability to provide multiple `--network` flags, hence I'd prefer to keep this one simple.

Reply:
Thanks for the digging! 

> * Always have the network "available" on every node (but it could be a "definition-only" for dynamically rolled-out networks)
> * Or, for attachable networks, and a container attaching to it, we could still consider to initialize it when attaching

Once this PR is merged, I'll submit a follow-up refactor PR to move attachment requests out of `findAndAttachNetwork` to a dedicated method called before doing anything else in `initializeNetworking`. With that change:

* On `ContainerCreate`, only soundness checks will be done ;
* On `ContainerStart` (and `NetworkConnect`), we'll determine whether referenced network exists locally, attach it if it's swarm scoped, or bail out because if it doesn't. And then, we can normalize and validate the settings thoroughly ;

I started doing some work in that direction here: https://github.com/moby/moby/commit/fba50d67d4a4c461c461732c2e1bc34e7e95eff5.

I think it'd be "conceptually" better to go back to the original semantic: _you can only attach to a network that exists locally_; but it'd be more work to implement and I guess we'd probably need to gate that behind an API version check, so anyway... I think we can get the same result by shuffling a bit the order of operations.

Reply:
> I think it'd be "conceptually" better to go back to the original semantic: you can only attach to a network that exists locally; but it'd be more work to implement and I guess we'd probably need to gate that behind an API version check, so anyway... I think we can get the same result by shuffling a bit the order of operations.

Yes, we need to look at that! I guess overall the tricky bit is that we must assume that there's potentially many nodes and many networks, and not all of those networks to be used on each node.

- For "attachable" networks, it would be good to have a ("stub?") network on each node, so that `docker network attache` can be used for containers; we could probably still make it a stub (so that it doesn't take up resources on _all notes_ in the swarm), and make it an actual network when attaching.
- For "non-attachable" networks, I guess it's TBD; do we want to show stubs for those? (if it's a _worker_ node, the worker itself wouldn't have much purpose for it until a swarm task lands on the node).
- For _managers_ it would be more useful (as you could run `docker network rm <some swarm-scope network>`)


Either way; all of that is for a follow-up discussion. We could create a ticket for it though (perhaps copy/paste some of the above)